presto-main/src/main/java/com/facebook/presto/operator/SpilledLookupSourceHandle.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.function.Supplier;
+
+import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.concurrent.MoreFutures.whenAnyComplete;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+final class SpilledLookupSourceHandle
+{
+    private enum State
+    {
+        SPILLED,
+        UNSPILLING,
+        PRODUCED,
+        DISPOSED
+    }
+
+    @GuardedBy("this")
+    private State state = State.SPILLED;
+
+    private final SettableFuture<?> unspillingRequested = SettableFuture.create();
+
+    @GuardedBy("this")
+    @Nullable
+    private SettableFuture<Supplier<LookupSource>> unspilledLookupSource;
+
+    private final SettableFuture<?> disposeRequested = SettableFuture.create();
+
+    private final ListenableFuture<?> unspillingOrDisposeRequested = whenAnyComplete(ImmutableList.of(unspillingRequested, disposeRequested));
+
+    public SettableFuture<?> getUnspillingRequested()
+    {
+        return unspillingRequested;
+    }
+
+    public synchronized ListenableFuture<Supplier<LookupSource>> getLookupSource()
+    {
+        assertIn(State.SPILLED);
+        unspillingRequested.set(null);
+        setState(State.UNSPILLING);
+        unspilledLookupSource = SettableFuture.create();
+        return unspilledLookupSource;
+    }
+
+    public synchronized void setLookupSource(Supplier<LookupSource> lookupSource)
+    {
+        requireNonNull(lookupSource, "lookupSource is null");
+
+        assertIn(State.UNSPILLING);
+        requireNonNull(unspilledLookupSource).set(lookupSource);
+        unspilledLookupSource = null; // let the memory go
+        setState(State.PRODUCED);
+    }
+
+    public synchronized void dispose()
```

Comment:
Is there a guarantee that the `setLookupSource` callback (looks like a callback), is coordinated with destroy?  If `setLookupSource` is called after destroy you will get an exception.  Instead you could ignore the call.